streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1531,58 +1540,121 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
+   */
+  final def scheduleElementsWith[R1 <: R, E1 >: E, B, C](
+    schedule: ZSchedule[R1, A, B]
+  )(f: A => C, g: B => C): ZStream[R1 with Clock, E1, C] =
+    ZStream[R1 with Clock, E1, C] {
+      for {
+        as    <- self.process
+        state <- Ref.make[Option[(A, schedule.State)]](None).toManaged_
+        pull = {
+          def go: Pull[R1 with Clock, E1, C] = state.get.flatMap {
+            case None =>
+              for {
+                a    <- as
+                init <- schedule.initial
+                _    <- state.set(Some(a -> init))
+              } yield f(a)
 
+            case Some((a, sched)) =>
+              for {
+                decision <- schedule.update(a, sched)
+                _        <- clock.sleep(decision.delay)
+                _        <- state.set(if (decision.cont) Some(a -> decision.state) else None)
+              } yield if (decision.cont) f(a) else g(decision.finish())
+          }
+
+          go
+        }
+      } yield pull
+    }
+
+  /**
+   * Uses the provided schedule to define the "rhythm" at which elements of the stream are emitted,
+   * finishing with the schedule's output (if the schedule is finite).
+   * Uses the provided function to align the stream and schedule outputs on the same type.
+   */
+  final def scheduleWith[R1 <: R, E1 >: E, B, C](
+    schedule: ZSchedule[R1, A, B]
+  )(f: A => C, g: B => C): ZStream[R1 with Clock, E1, C] =
+    ZStream[R1 with Clock, E1, C] {
       for {
         as    <- self.process
         init  <- schedule.initial.toManaged_
-        state <- Ref.make[State](State.Initial(init)).toManaged_
+        state <- Ref.make[(Boolean, schedule.State, Option[() => B])]((false, init, None)).toManaged_
         pull = state.get.flatMap {
-          case State.Initial(sched) =>
-            for {
-              a        <- as
-              decision <- schedule.update(a, sched)
-              _        <- state.set(decide(decision, a))
-            } yield Right(a)
-
-          case State.SleepAndRepeat(sched, delay, a) =>
-            for {
-              _        <- clock.sleep(delay)
-              decision <- schedule.update(a, sched)
-              _        <- state.set(decide(decision, a))
-            } yield Right(a)
-
-          case State.SleepAndFinish(delay, b) =>
-            for {
-              _ <- clock.sleep(delay)
-              _ <- state.set(State.Initial(init))
-            } yield Left(b)
+          case (done, sched, decision0) =>
+            if (done) Pull.end
+            else
+              for {
+                a <- as.optional.mapError(Some(_))
+                c <- a match {
+                      case Some(a) =>
+                        for {
+                          decision <- schedule.update(a, sched)
+                          _        <- clock.sleep(decision.delay)
+                          _        <- state.set((!decision.cont, decision.state, Some(decision.finish)))
```

Comment:
Hmm didn't we say that the schedule should not stop the stream?

Reply:
Ah I see that my example was unclear. In this formulation, I meant that the schedule should restart between the elements. So for a stream `"A B C A B C"`, this should be result:
```
A |-2ms-| B |-4ms-| C |-8ms-| "Done" A |-2ms-| B |-4ms-| C |-8ms-| "Done"
```

Reply:
Nope, this is implement for point 2 above:

> 2. As long as the schedule wants to continue, emit the stream's next element after a delay

Reply:
And the short circuiting can be attained by `collectWhile { case Right(v) => v }`, which would stop on the first `Left`.

Reply:
(sorry for dragging this!)

Reply:
Actually, let's merge this and address in a follow-up.

Reply:
That is the semantics of `scheduleElements`, isn't it? 🤔 I could add a test to prove that.

Reply:
Ah not quite. `scheduleElements` repeats every element and restarts